base/operators.jl

```diff
@@ -224,16 +224,44 @@ function promote_shape(a::Dims, b::Dims)
     return a
 end
 
+# The lengths of the given indices, lowering : to the appropriate size
+stagedfunction index_lengths(A::AbstractArray, I...)
+    N = length(I)
+    ex = Expr(:tuple)
+    for d=1:N
+        if I[d] <: Colon
+            push!(ex.args, d < N  ? (:(size(A, $d))) :
+                           d == 1 ? (:(length(A)))   : (:(trailingsize(A, $d))))
+        elseif I[d] <: Real
+            push!(ex.args, 1)
+        else
+            push!(ex.args, :(length(I[$d])))
+        end
+    end
+    ex
+end
+
 # shape of array to create for getindex() with indexes I
 # drop dimensions indexed with trailing scalars
-index_shape(I::Real...) = ()
-index_shape(i, I...) = tuple(length(i), index_shape(I...)...)
+stagedfunction index_shape(A::AbstractArray, I...)
```

Comment:
Hm, I'll have to think about that.  It's not as easy to recursively pick off the elements with `:` in the mix because the colon needs to know its position in the argument list.